src/mapeo-manager.js

```diff
@@ -0,0 +1,139 @@
+import { randomBytes } from 'crypto'
+import path from 'path'
+import { KeyManager } from '@mapeo/crypto'
+import Database from 'better-sqlite3'
+import { eq } from 'drizzle-orm'
+import { drizzle } from 'drizzle-orm/better-sqlite3'
+import { migrate } from 'drizzle-orm/better-sqlite3/migrator'
+import { IndexWriter } from './index-writer/index.js'
+import { MapeoProject } from './mapeo-project.js'
+import { projectKeysTable, projectTable } from './schema/client.js'
+import { ProjectKeys } from './generated/keys.js'
+
+const CLIENT_SQLITE_FILE_NAME = 'client.db'
+
+export class MapeoManager {
+  #keyManager
+  #projectSettingsIndexWriter
+  #storagePath
+  #db
+
+  /**
+   * @param {Object} [opts]
+   * @param {string} [opts.storagePath] Folder for all data storage (hypercores and sqlite db). Folder must exist. If not defined, everything is stored in-memory
+   */
+  constructor({ storagePath } = {}) {
+    const dbPath =
+      storagePath !== undefined
+        ? path.join(storagePath, CLIENT_SQLITE_FILE_NAME)
+        : ':memory:'
+
+    const sqlite = new Database(dbPath)
+    this.#db = drizzle(sqlite)
+    migrate(this.#db, { migrationsFolder: './drizzle/client' })
+
+    this.#keyManager = new KeyManager(KeyManager.generateRootKey())
+    this.#projectSettingsIndexWriter = new IndexWriter({
+      tables: [projectTable],
+      sqlite,
+    })
+    this.#storagePath = storagePath
+  }
+
+  /**
+   * Create a new project.
+   * @param {import('type-fest').Simplify<Partial<Pick<import("@mapeo/schema").ProjectValue, 'name'>>>} settings
+   * @returns {Promise<string>}
+   */
+  async createProject(settings) {
+    // 1. Create project keypair
+    const projectKeypair = KeyManager.generateProjectKeypair()
+
+    // 2. Create namespace encryption keys
+    /** @type {Record<import('./core-manager/core-index.js').Namespace, Buffer>} */
+    const encryptionKeys = {
+      auth: randomBytes(32),
+      blob: randomBytes(32),
+      blobIndex: randomBytes(32),
+      config: randomBytes(32),
+      data: randomBytes(32),
+    }
+
+    // 3. Save keys to client db in projectKeys table
+    /** @type {import('./types.js').ProjectKeys} */
+    const keys = {
+      projectKey: projectKeypair.publicKey,
+      projectSecretKey: projectKeypair.secretKey,
+      encryptionKeys,
+    }
+
+    // TODO: Update to use @mapeo/crypto when ready (https://github.com/digidem/mapeo-core-next/issues/171)
+    const projectId = projectKeypair.publicKey.toString('hex')
+
+    this.#db
+      .insert(projectKeysTable)
+      .values({
+        projectId,
+        keysCipher: this.#keyManager.encryptLocalMessage(
+          Buffer.from(ProjectKeys.encode(keys).finish()),
```

Comment:
is this the correct way of doing this? seems a little verbose but needed for TS reasons (`finish` returns `Uint8Array` but `encryptLocalMessage` accepts `Buffer`)

Reply:
Ah I should update the key manager method to accept uint8array. Fine to TS ignore for now. At some point need to go through the code and use b4a.

Reply:
Issue created: https://github.com/digidem/mapeo-crypto/issues/14

Reply:
simplified based on your suggestion and added a todo comment via https://github.com/digidem/mapeo-core-next/pull/199/commits/b3634e6b3b45e0de5ffbb7bf6ce3f82e0c177caf